src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1606,204 @@ class OptimizeShuffles : public IRMutator {
+                return shifted_load;
+            } else if (maybe_shuffle->is_concat()) {
+                return are_contiguous_vectors(maybe_shuffle->vectors);
+            }
+        }
+        if (const Load *maybe_load = e.as<Load>()) {
+            const Ramp *maybe_ramp = maybe_load->index.as<Ramp>();
+            if (maybe_ramp && is_const(maybe_ramp->stride, 1)) {
+                return maybe_load;
+            }
+        }
+        return Expr();
+    }
+
+    // Loads comparator for sorting Load Expr of the same buffer.
+    static bool loads_comparator(LoadIndex a, LoadIndex b) {
+        if (a.first.defined() && b.first.defined()) {
+            const Load* load_a = a.first.as<Load>();
+            const Load* load_b = b.first.as<Load>();
+            if (load_a->name == load_b->name) {
+                Expr base_diff = simplify(load_b->index - load_a->index);
+                if (is_positive_const(base_diff)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    // Vtmpy helps in sliding window ops of the form a*v0 + b*v1 + v2.
+    // Conditions required:
+    //      v0, v1 and v2 start indices differ by vector stride
+    // Current supported value of stride is 1.
+    // TODO: Add support for any stride.
+    void visit(const Add *op) {
+        // Find opportunities vtmpy
+        if (op && op->type.is_vector() && (op->type.bits() == 16 || op->type.bits() == 32)) {
+            int lanes = op->type.lanes();
+            vector<MulExpr> mpys;
+            Expr rest;
+            string vtmpy_suffix;
+
+            // Finding more than 100 such expresssions is rare.
+            // Setting it to 100 makes sure we dont miss anything
+            // in most cases and also dont spend unreasonable time while
+            // just looking for vtmpy patterns.
+            if (op->type.bits() == 16) {
+                find_mpy_ops(op, UInt(8, lanes), Int(8), 100, mpys, rest);
+                vtmpy_suffix = ".vub.vub.b.b";
+                if (mpys.size() < 3) {
+                    mpys.clear();
+                    rest = Expr();
+                    find_mpy_ops(op, Int(8, lanes), Int(8), 100, mpys, rest);
+                    vtmpy_suffix = ".vb.vb.b.b";
+                }
+            } else if (op->type.bits() == 32) {
+                find_mpy_ops(op, Int(16, lanes), Int(8), 100, mpys, rest);
+                vtmpy_suffix = ".vh.vh.b.b";
+            }
+
+            if (mpys.size() >= 3) {
+                const size_t mpy_size = mpys.size();
+                // Used to put loads with different buffers in different buckets.
+                std::unordered_map<string, vector<LoadIndex> > loads;
+                // To keep track of indices selected for vtmpy.
+                std::unordered_map<size_t, bool> vtmpy_indices;
+                vector<Expr> vtmpy_exprs;
+                const char *not_load = "";
```

Comment:
Looking at this not_load code... could you just start the 'new_expr' here, and when you find something that isn't a load, just add it to the new_expr eagerly, instead of putting it in the map of loads? Then you wouldn't need this not_load token. It's up to you, use your judgement if this makes the code reasonably simpler or not.